metadata-io/src/main/java/com/linkedin/metadata/entity/EntityServiceImpl.java

```diff
@@ -527,227 +481,192 @@ public ListResult<RecordTemplate> listLatestAspects(
    */
   @Nonnull
-  @Deprecated
-  protected UpdateAspectResult patchAspectToLocalDB(
-      @Nonnull final Urn urn,
-      @Nonnull final AspectSpec aspectSpec,
-      @Nonnull final Patch jsonPatch,
-      @Nonnull final AuditStamp auditStamp,
-      @Nonnull final SystemMetadata providedSystemMetadata) {
+  private List<UpdateAspectResult> ingestAspectsToLocalDB(@Nonnull final AspectsBatch aspectsBatch,
+                                                          @Nonnull final AuditStamp auditStamp,
+                                                          boolean overwrite) {
 
-    return _aspectDao.runInTransactionWithRetry(() -> {
-      final String urnStr = urn.toString();
-      final String aspectName = aspectSpec.getName();
-      final EntityAspect latest = _aspectDao.getLatestAspect(urnStr, aspectName);
-      final long nextVersion = _aspectDao.getNextVersion(urnStr, aspectName);
-      try {
-
-        final RecordTemplate currentValue = latest != null
-            ? EntityUtils.toAspectRecord(urn, aspectName, latest.getMetadata(), _entityRegistry)
-            : _entityRegistry.getAspectTemplateEngine().getDefaultTemplate(aspectSpec.getName());
-
-        if (latest == null && currentValue == null) {
-          // Attempting to patch a value to an aspect which has no default value and no existing value.
-          throw new UnsupportedOperationException(String.format("Patch not supported for aspect with name %s. "
-              + "Default aspect is required because no aspect currently exists for urn %s.", aspectName, urn));
-        }
-
-        final RecordTemplate updatedValue = _entityRegistry.getAspectTemplateEngine().applyPatch(currentValue, jsonPatch, aspectSpec);
+    if (aspectsBatch.containsDuplicateAspects()) {
+      log.warn(String.format("Batch contains duplicates: %s", aspectsBatch));
+    }
 
-        validateAspect(urn, updatedValue);
-        return ingestAspectToLocalDBNoTransaction(urn, aspectName, ignored -> updatedValue, auditStamp, providedSystemMetadata,
-            latest, nextVersion);
-      } catch (JsonProcessingException | JsonPatchException e) {
-        throw new IllegalStateException(e);
+    return _aspectDao.runInTransactionWithRetry((tx) -> {
+      // Read before write is unfortunate, however batch it
+      Map<String, Set<String>> urnAspects = aspectsBatch.getUrnAspectsMap();
+      // read #1
+      Map<String, Map<String, EntityAspect>> latestAspects = _aspectDao.getLatestAspects(urnAspects);
+      // read #2
+      Map<String, Map<String, Long>> nextVersions = _aspectDao.getNextVersions(urnAspects);
+
+      List<UpsertBatchItem> items = aspectsBatch.getItems().stream()
+              .map(item -> {
+                if (item instanceof UpsertBatchItem) {
+                  return (UpsertBatchItem) item;
+                } else {
+                  // patch to upsert
+                  PatchBatchItem patchBatchItem = (PatchBatchItem) item;
```

Comment:
Merged the paths for PATCH and Upsert, they were practically identical except we take a patch and apply it to the latest aspect. Essentially patching just converts it from a patch item to a normal upsert item.